File: samples_c/s110_queue_module/queue.c
#include "queue.h"

void queue_init(Queue *q) {
    q->front = 0;
    q->rear = 0;
    q->count = 0;
}

int queue_is_empty(const Queue *q) {
    return q->count == 0;
}

int queue_is_full(const Queue *q) {
    return q->count == QUEUE_CAPACITY;
}

int queue_enqueue(Queue *q, int value) {
    if (queue_is_full(q)) return 0;
    q->data[q->rear] = value;
    q->rear = (q->rear + 1) % QUEUE_CAPACITY;
    q->count++;
    return 1;
}

int queue_dequeue(Queue *q, int *out) {
    if (queue_is_empty(q)) return 0;
    *out = q->data[q->front];
    q->front = (q->front + 1) % QUEUE_CAPACITY;
    q->count--;
    return 1;
}

int queue_size(const Queue *q) {
    return q->count;
}

File: translations_rust_haiku/s110_queue_module/queue.rs
pub const QUEUE_CAPACITY: usize = 8;

pub struct Queue {
    pub data: [i32; QUEUE_CAPACITY],
    pub front: usize,
    pub rear: usize,
    pub count: usize,
}

impl Queue {
    pub fn new() -> Self {
        Queue {
            data: [0; QUEUE_CAPACITY],
            front: 0,
            rear: 0,
            count: 0,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.count == 0
    }

    pub fn is_full(&self) -> bool {
        self.count == QUEUE_CAPACITY
    }

    pub fn enqueue(&mut self, value: i32) -> bool {
        if self.is_full() {
            return false;
        }
        self.data[self.rear] = value;
        self.rear = (self.rear + 1) % QUEUE_CAPACITY;
        self.count += 1;
        true
    }

    pub fn dequeue(&mut self) -> Option<i32> {
        if self.is_empty() {
            return None;
        }
        let value = self.data[self.front];
        self.front = (self.front + 1) % QUEUE_CAPACITY;
        self.count -= 1;
        Some(value)
    }

    pub fn size(&self) -> usize {
        self.count
    }
}
